packages/devtools_app/lib/src/timeline/timeline_model.dart

```diff
@@ -506,6 +624,196 @@ class TimelineEvent extends TreeNode<TimelineEvent> {
   }
 }
 
+class SyncTimelineEvent extends TimelineEvent {
+  SyncTimelineEvent(TraceEventWrapper firstTraceEvent) : super(firstTraceEvent);
+
+  bool get isUiEventFlow => containsChildWithCondition(
+      (TimelineEvent event) => event.name.contains('Engine::BeginFrame'));
+
+  bool get isGpuEventFlow => containsChildWithCondition(
+      (TimelineEvent event) => event.name.contains('PipelineConsume'));
+
+  @override
+  int get displayDepth => depth;
+
+  @override
+  bool get hasOverlappingChildren => false;
+
+  @override
+  bool couldBeParentOf(TimelineEvent e) {
+    final startTime = time.start.inMicroseconds;
+    final endTime = time.end?.inMicroseconds;
+    final eStartTime = e.time.start.inMicroseconds;
+    final eEndTime = e.time.end?.inMicroseconds;
+
+    if (endTime != null && eEndTime != null) {
+      if (startTime == eStartTime && endTime == eEndTime) {
+        return traceEvents.first.id < e.traceEvents.first.id;
+      }
+      return startTime <= eStartTime && endTime >= eEndTime;
+    } else if (endTime != null) {
+      // We don't use >= to compare [endTime] and [e.startTime] here because we
+      // don't want to falsely make [this] the parent of [e]. We do not know
+      // [e.endTime], meaning [e] could start at [endTime] and end later than
+      // [endTime] (unless e has a duration of 0). In this case, [this] would
+      // not be the parent of [e].
+      return startTime <= eStartTime && endTime > eStartTime;
+    } else if (startTime == eStartTime) {
+      return traceEvents.first.id < e.traceEvents.first.id;
+    } else {
+      return startTime < eStartTime;
+    }
+  }
+}
+
+class AsyncTimelineEvent extends TimelineEvent {
```

Comment:
how much of this logic is duplicated with SyncTimelineEvent?

Reply:
None of the logic is duplicated. They both implement abstract methods from their superclass.